File: sykli.go
//go:build ignore

// sykli.go - CI pipeline for Tapio
//
// Local:  sykli
// CI:     GitHub Actions runs sykli
package main

import sykli "github.com/yairfalse/sykli/sdk/go"

func main() {
	s := sykli.New()

	// === STATIC ANALYSIS (parallel) ===
	s.Task("fmt-check").
		Run("gofmt -l . | grep -q . && exit 1 || exit 0").
		Inputs("**/*.go")

	s.Task("vet").
		Run("go vet ./...").
		Inputs("**/*.go", "go.mod", "go.sum")

	s.Task("lint").
		Run("golangci-lint run --timeout 5m").
		Inputs("**/*.go", "go.mod", "go.sum")

	// === TESTS (after static analysis) ===
	s.Task("test").
		Run("go test -timeout 5m ./...").
		After("fmt-check", "vet", "lint").
		Inputs("**/*.go", "go.mod", "go.sum")

	// === BUILD (after tests) ===
	s.Task("build").
		Run("mkdir -p bin && CGO_ENABLED=0 go build -ldflags='-s -w' -o bin/tapio ./cmd/tapio 2>/dev/null || echo 'No main.go yet'").
		After("test")

	s.Emit()
}

File: sykli.rs
#!/usr/bin/env -S cargo +nightly -Zscript
---
[dependencies]
sykli = { git = "https://github.com/yairfalse/sykli", package = "sykli-sdk-rust" }
---

fn main() {
    let s = sykli::Pipeline::new();

    // Static analysis (parallel)
    s.task("fmt-check")
        .run("cargo fmt --check")
        .inputs(&["**/*.rs", "Cargo.toml", "Cargo.lock"]);

    s.task("clippy")
        .run("cargo clippy --workspace --all-targets -- -D warnings")
        .inputs(&["**/*.rs", "Cargo.toml", "Cargo.lock"]);

    // Keep controller/server/Kubernetes deps out of the node binary.
    s.task("agent-deps")
        .run("scripts/check-agent-deps.sh")
        .inputs(&["Cargo.toml", "Cargo.lock", "scripts/check-agent-deps.sh"]);

    // Tests (after static analysis)
    s.task("test")
        .run("cargo test --workspace")
        .after(&["fmt-check", "clippy", "agent-deps"])
        .inputs(&["**/*.rs", "Cargo.toml", "Cargo.lock"]);

    // Build (after tests)
    s.task("build")
        .run("cargo build --release")
        .after(&["test"]);

    s.emit();
}
